instrumentation/http-client/src/OpenTelemetry/Instrumentation/HttpClient.hs

```diff
@@ -13,143 +15,144 @@
-    pure resp
+import Control.Exception (assert)
+import Control.Monad ((>=>))
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Data.Bifunctor (bimap)
+import qualified Data.CaseInsensitive as CI
+import Data.Foldable (fold)
+import Data.Function ((&))
+import qualified Data.HashMap.Strict as H
+import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.Int (Int64)
+import Data.Monoid (Endo (Endo, appEndo))
+import qualified Data.TLS.GHC as TLS
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import GHC.Stack (HasCallStack, withFrozenCallStack)
+import qualified Network.HTTP.Client as Orig
+import qualified Network.HTTP.Types.Status as HT
+import qualified OpenTelemetry.Attributes.Map as Otel
+import qualified OpenTelemetry.Context.ThreadLocal as Otel
+import qualified OpenTelemetry.Propagator as Otel
+import qualified OpenTelemetry.SemanticConventions as Otel
+import qualified OpenTelemetry.Trace.Core as Otel
+
+
+appendModifierToSettings :: (MonadIO m, HasCallStack) => Otel.TracerProvider -> Orig.ManagerSettings -> m Orig.ManagerSettings
+appendModifierToSettings tracerProvider settings = withFrozenCallStack $ liftIO $ do
+  let
+    tracer =
+      Otel.makeTracer
+        tracerProvider
+        $Otel.detectInstrumentationLibrary
+        Otel.tracerOptions
+    managerModifyRequest = Orig.managerModifyRequest settings
+    managerModifyResponse = Orig.managerModifyResponse settings
+  tls <- makeThreadLocalStorage
+  pure
+    settings
+      { Orig.managerModifyRequest = requestModifier tracer tls >=> managerModifyRequest
+      , Orig.managerModifyResponse = managerModifyResponse >=> responseModifier tls
+      }
+
+
+type ThreadLocalStorage = TLS.TLS (IORef (Maybe Otel.Span))
+
+
+requestModifier :: HasCallStack => Otel.Tracer -> ThreadLocalStorage -> Orig.Request -> IO Orig.Request
+requestModifier tracer tls request = do
+  spanRef <- TLS.getTLS tls
+  maybeSpan <- readIORef spanRef
+  case maybeSpan of
+    Nothing -> do
+      context <- Otel.getContext
+      let attributes = makeRequestAttributes request `H.union` Otel.callerAttributes
+      span_ <- Otel.createSpan tracer context "request" Otel.defaultSpanArguments {Otel.kind = Otel.Client, Otel.attributes}
+      writeIORef spanRef $ Just span_
+      let propagator = Otel.getTracerProviderPropagators $ Otel.getTracerTracerProvider tracer
+      headers <- Otel.injector propagator context $ Orig.requestHeaders request
+      pure request {Orig.requestHeaders = headers}
+    Just _ -> pure request
+
+
+responseModifier :: HasCallStack => ThreadLocalStorage -> Orig.Response Orig.BodyReader -> IO (Orig.Response Orig.BodyReader)
+responseModifier tls response = do
+  spanRef <- TLS.getTLS tls
+  maybeSpan <- readIORef spanRef
+  case maybeSpan of
+    Just span_ -> do
+      Otel.addAttributes span_ $ makeResponseAttributes response
+      Otel.endSpan span_ Nothing
```

Comment:
What happens if there is no response? or if an exception is thrown? I think this version is less safe in the presence of failures than the old one. The old one used the `inSpan` functions, so should never leave us with unterminated spans, but I think this one can do?